internal/core/task.go

```diff
@@ -0,0 +1,280 @@
+		return err
+	}
+	url := fmt.Sprintf("%s%s", c.domain, urlKey)
+
+	resp, err := c.translation.Engine.Video2Text(url, "")
+	if err != nil {
+		return err
+	}
+
+	err = c.createVideoTask(ctx, userId, resourceId, resp.TaskId)
+	if err != nil {
+		return err
+	}
+
+	// Set video task cache
+	c.SetVideoTaskCache(resourceId, VideoTaskTimestamp(time.Now().Unix()))
+
+	c.xLog.Infof("NewVideoTask success, resourceId: %s, taskId: %s", resourceId, resp.TaskId)
+
+	return nil
+}
+
+// SetVideoTaskSuccess set video task success
+func (c *Community) SetVideoTaskSuccess(ctx context.Context, resourceId string) error {
+	return c.updateVideoTaskStatus(ctx, resourceId, 1)
+}
+
+// SetVideoTaskFailed set video task failed
+func (c *Community) SetVideoTaskFailed(ctx context.Context, resourceId string) error {
+	return c.updateVideoTaskStatus(ctx, resourceId, -1)
+}
+
+// SetVideoTaskOutput set video task output link
+func (c *Community) SetVideoTaskOutput(ctx context.Context, resourceId string, output string) error {
+	return c.updateVideoTaskOutput(ctx, resourceId, output)
+}
+
+type VideoTaskTimestamp int64
+type VideoTaskMap map[string]VideoTaskTimestamp
+
+// VideoTaskCache video task cache
+// Simple cache, no expiration
+type VideoTaskCache struct {
+	sync.RWMutex
+	videoTaskMap VideoTaskMap
+
+	// Check task status
+	isActive bool
+}
+
+// NewVideoTaskCache create new video task cache
+func NewVideoTaskCache() *VideoTaskCache {
+	return &VideoTaskCache{
+		videoTaskMap: make(VideoTaskMap),
+		isActive:     false,
+	}
+}
+
+// Get get video task timestamp
+func (c *VideoTaskCache) Get(key string) (VideoTaskTimestamp, bool) {
+	c.RLock()
+	defer c.RUnlock()
+	value, ok := c.videoTaskMap[key]
+	return value, ok
+}
+
+// Set set video task timestamp
+func (c *VideoTaskCache) Set(key string, value VideoTaskTimestamp) {
+	c.Lock()
+	defer c.Unlock()
+	c.videoTaskMap[key] = value
+}
+
+// Delete delete video task
+func (c *VideoTaskCache) Delete(key string) {
+	c.Lock()
+	defer c.Unlock()
+	delete(c.videoTaskMap, key)
+}
+
+// Clear clear video task cache
+func (c *VideoTaskCache) Clear() {
+	c.Lock()
+	defer c.Unlock()
+	c.videoTaskMap = make(VideoTaskMap)
+}
+
+// SetVideoTaskCache set video task cache
+func (c *Community) SetVideoTaskCache(key string, value VideoTaskTimestamp) {
+	c.translation.VideoTaskCache.Set(key, value)
+
+	// Start check task status
+	if !c.translation.VideoTaskCache.isActive {
+		c.translation.VideoTaskCache.isActive = true
+		go c.TimedCheckVideoTask(context.Background(), 30*time.Second)
+	}
+}
+
+// Timed check status of video task with timeout
+func (c *Community) TimedCheckVideoTask(ctx context.Context, timeout time.Duration) {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.xLog.Infof("TimedCheckVideoTask start, timeout: %v", timeout)
+			// Iterator video task cache
+			c.translation.VideoTaskCache.RLock()
```

Comment:
this lock seems too big

Reply:
This issue has been solved.
<img width="1049" alt="image" src="https://github.com/goplus/community/assets/47499836/1ee84b47-6d7a-456a-9c42-6da6ef8ba631">